# SARS-COV-2_B4B
This is a repository for development of the SARS-CoV-2 Bioinformatics for Beginners Course

SARS-CoV-2 variant lineage identification is key to pandemic tracking and enabling public health response. This course is an introduction to bioinformatics by applying skills used in SARS-CoV-2 genomic data analysis. This will be a distributed classrooms style course run across Africa; Latin America and the Caribbean; and Asia. This model was developed by [H3ABioNet](https://www.h3abionet.org), see this [publication](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008640) for more info. 

SARS-CoV-2 variant lineage identification is key to pandemic tracking and enabling public health response. This course is an introduction to bioinformatics by applying skills used in SARS-CoV-2 genomic data analysis. Bioinformatics skills are fundamental in management and assessment of viral sequences. This course will introduce you to processing data programmatically, the data formats used in viral sequencing, how to determine the variant lineage (Delta, Omicron etc.), and how to share data so that others around the world can benefit. These skills are the building blocks for scaling up analysis to pandemic response levels. 

[Course website](https://coursesandconferences.wellcomeconnectingscience.org/event/sars-cov-2-bioinformatics-for-beginners-20221031/)

## Course structure 
This course is making use of Google Colab - https://colab.research.google.com/, a free to use service.

Access to Colab is via a [Google Account](https://www.google.com/account/about/), which can be made for free.

## Time commitment
Contact sessions will run twice a week,  lasting for 4 hours per session. It will run between the 31st of October â€“ 2nd of December 2022. There will be sessions in two time zones. Note, each session for Oceania and Asia; and Latin America and Africa; will run in the same block of time, but with regional time differences. 

## Target audience
The course is aimed at postgraduate scientists, postdoctoral scientists,  junior faculty members or clinicians/healthcare professionals based in the regions across Africa, Asia, and Latin America & the Caribbean. It does not require bioinformatics skills as a prerequisite.

## Programme
The programme will cover the following core topics:

- Intro to Python Notebooks
- Intro to Unix/Linux & running commands
- Introduction to NGS Technologies employed in SARS-CoV-2 sequencing
- Data quality control
- Workflows for sequencing analysis
- Pangolin for lineage identification
- Exploring genomics data in a global context

## Learning Outcomes

- Apply command line tools for sequence data quality control
- List file formats commonly used in SARS-CoV-2 sequencing
- Use Pangolin to create viral lineages from sets of existing data
- List key metadata that must be included when uploading sequences to online repositories
- Describe broad principles in translation of analysis outputs to outbreak/epi/pandemic response

## Instructor Team
<!-- example - [Carolina Torres](https://www.researchgate.net/profile/Carolina-Torres-7), Universidad de Buenos Aires - CONICET, Buenos Aires, Argentina   -->
- [Ariel Amadio](https://www.unraf.edu.ar/), Universidad Nacional de Rafaela, IDICAL, [INTA-CONICET](https://idical.conicet.gov.ar/), Argentina
- [Idowu Olawoye](link)
- [John Juma](link) 
- [Jorge Batista da Rocha](https://coursesandconferences.wellcomeconnectingscience.org/about-us/the-team/), Wellcome Connecting Science, United Kingdom
- [Leigh Jackson](link)
- [Marcela Suarez Esquivel](link), Universidad Nacional, Costa Rica <!-- no permission to share institution link -->
- [Nicholas Bbosa](link)
- [Paul Oluniyi](link)
- [Progress Dube](link)
- [Una Ren](https://www.esr.cri.nz/), Special Pathogens Unit, Environmental Science and Research, New Zeland
- [Zahra Waheed](link)

## Timetable
[View the timetable here](https://docs.google.com/document/d/15LkC6fEp2vQfZKm8KfkLyPb8bRpqfjusM-fLWa7Kic4/edit?usp=sharing)

## Course manual

**Introduction**       
[Get to know your classroom and assistants](link)       
[Learn how to use moodle and make a classroom bio](link)      

**Module 1: Introduction to Notebooks & Unix command line**      
[Notebooks and Unix - Part 1](https://github.com/WCSCourses/SARS-COV-2_B4B/blob/main/Modules/Module_1_readme.md)     
[Notebooks and Unix - Part 2](https://github.com/WCSCourses/SARS-COV-2_B4B/blob/main/Modules/Module_1_readme.md)      

**Module 2: Data QC and Consensus sequences**    
[Data QC and Consensus Part 1](https://github.com/WCSCourses/SARS-COV-2_B4B/blob/main/Modules/Module_2_readme.md)       
[Data QC and Consensus Part 2](https://github.com/WCSCourses/SARS-COV-2_B4B/blob/main/Modules/Module_2_readme.md)     

**Module 3: Variant Lineage Identification**       
[Variant Lineage Identification Part 1](https://github.com/WCSCourses/SARS-COV-2_B4B/blob/main/Modules/Module_3_readme.md)      
[Variant Lineage Identification Part 2](https://github.com/WCSCourses/SARS-COV-2_B4B/blob/main/Modules/Module_3_readme.md)       

**Module 4: Data sharing and interpretation**       
[Data sharing and interpretation Part 1](https://github.com/WCSCourses/SARS-COV-2_B4B/blob/main/Modules/Module_4_readme.md)      
[Data sharing and interpretation Part 2](https://github.com/WCSCourses/SARS-COV-2_B4B/blob/main/Modules/Module_4_readme.md)       

Any reuse of the course materials, data or code is encouraged with due acknowledgement.

******
## License
<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
